lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUTransformCharFilter.java

```diff
@@ -0,0 +1,384 @@
+            preLimit = bufferLength;
+          }
+          cursorAdvanced(preStart, preLimit);
+        }
+      }
+    }
+
+    return -1;
+  }
+
+  private void rollback(int preStart, int preLimit) {
+    buffer.delete(preStart, position.limit); // delete uncommitted chars
+    buffer.insert(preStart, rollbackBuffer, 0, rollbackBufferSize);
+    position.start = preStart;
+    position.contextLimit = preLimit;
+    position.limit = preLimit;
+  }
+
+  /**
+   * Grow rollback buffer if necessary, within constraints of {@link #maxRollbackBufferCapacity}.
+   * This imposes an arbitrary failsafe to prevent the possibility that the rollback buffer could grow
+   * indefinitely.
+   *
+   * @return true if upon return, rollback buffer has sufficient capacity for new input chars, otherwise false.
+   */
+  private boolean ensureRollbackBufferCapacity() {
+    // ensure space for at least 2 chars (surrogate pair, max possible space needed)
+    if (rollbackBuffer.length - rollbackBufferSize < 2) {
+      if (rollbackBuffer.length < maxRollbackBufferCapacity) {
+        rollbackBuffer = ArrayUtil.growExact(rollbackBuffer, rollbackBuffer.length << 1);
+      } else {
+        // hit threshold; not going to increase the buffer size
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private int pushRollbackBuffer(int idx, int bufferLength) {
+    if (idx >= bufferLength) {
+      return 0;
+    } else {
+      // note: we have already ensured sufficient buffer capacity
+      final char candidate = buffer.charAt(idx);
+      if (!UTF16.isLeadSurrogate(candidate)) {
+        if (rollbackBuffer != null) {
+          rollbackBuffer[rollbackBufferSize++] = candidate;
+        }
+        return 1;
+      } else {
+        if (++idx < bufferLength) {
+          if (rollbackBuffer != null) {
+            rollbackBuffer[rollbackBufferSize++] = candidate;
+            rollbackBuffer[rollbackBufferSize++] = buffer.charAt(idx);
+          }
+          return 2;
+        } else {
+          // we don't yet have the high surrogate
+          if (inputFinished) {
+            // no more input available; proceed with the char we have; rollback not needed
+            return 1;
+          } else {
+            return 0; // wait for more input to be available
+          }
+        }
+      }
+    }
+  }
+
+  private int readFromIoNormalizeUptoBoundary() {
+    int nextCharLength = pushRollbackBuffer(position.limit, buffer.length());
+    if (nextCharLength == 0) {
+      return 0;
+    }
+    final int preCharCount = charCount;
+    int preStart = position.start;
+    int preLimit;
+    do {
+      position.limit += nextCharLength;
+      preLimit = position.limit;
+      position.contextLimit = preLimit;
+      transform.filteredTransliterate(replaceable, position, true);
+      boolean rollbackSizeWithinBounds = true;
+      if (rollbackBuffer != null && position.start < position.limit && (rollbackSizeWithinBounds = ensureRollbackBufferCapacity())) {
```

Comment:
Is it possible to move the side effect (assignment to rollbackSizeWIthinBounds) inside this if statement, and not to do it as part of the test expression? I think if you use a `continue` in this block you can do it?

Reply:
I can't think of a straightforward way to do this. Basically, we only need to `ensureRollbackBufferCapacity` if we're going to use the buffer to rollback (i.e., if the first two conditional statements are true). And then in the `else` block, we need to know whether we got there because `position.start == position.limit` (business as usual), or because we _wanted_ to rollback but there was no space in the rollback buffer.
One option would be to `ensureRollbackBufferCapacity()` _every_ time, but for the many Transliterators that will never need to _use_ the rollback buffer, this extra work would be totally unnecessary (and probably not JIT-optimizable in a runtime with several different active Transliterators?).